routes/v1.js

```diff
@@ -335,6 +358,141 @@ api.declare({
 });
 
 api.declare({
+  method: 'put',
+  route: '/secret/:token',
+  name: 'createSecret',
+  scopes: ['aws-provisioner:create-secret'],
+  input: SCHEMA_PREFIX_CONST + 'create-secret-request.json#',
+  title: 'Create new Secret',
+  description: [
+    'Insert a secret into the secret storage.  This should not',
+    'normally be done through this API, but is provided for testing',
+    'and completeness',
+  ].join('\n'),
+}, async function (req, res) {
+  var input = req.body;
+  var token = req.params.token;
+
+  var secret;
+  try {
+    secret = await this.Secret.create({
+      token: token,
+      workerType: input.workerType,
+      secrets: input.secrets,
+      scopes: input.scopes,
+      expiration: new Date(input.expiration),
+    });
+  } catch (err) {
+    if (err.code !== 'EntityAlreadyExists') {
+      throw err;
+    }
+
+    secret = await this.Secret.load({
+      token: token,
+      provisionerId: this.provisionerId,
+    });
+
+    var match = [
+      'workerType',
+      'secrets',
+      'token',
+      'scopes',
+      //'expiration' weird stuff is happening here.  going to assume that
+      // we should probably do some sort of Date.toISOString() comparison or something
+    ].every((key) => {
+      return _.isEqual(secret[key], input[key]);
+    });
+
+    // If we don't have a match we return 409, otherwise we continue as this is
+    // is an idempotent operation.
+    if (!match) {
+      res.status(409).json({
+        error: 'Secret already exists with different definition',
+      });
+      return;
+    }
+  }
+
+  res.reply({outcome: 'success'});
+  return;
+});
+
+api.declare({
+  method: 'get',
+  route: '/secret/:token',
+  name: 'getSecret',
+  output: SCHEMA_PREFIX_CONST + 'get-secret-response.json#',
+  title: 'Get a Secret',
+  description: [
+    'Retreive a secret from storage.  It is important that this secret is',
+    'deleted by the consumer, or else the secrets will be visible to any',
+    'process which can read HTTP on the worker localhost interface.',
+  ].join('\n'),
+}, async function (req, res) {
+  var token = req.params.token;
+  var that = this;
+
+  var p = this.Secret.load({
+    token: token,
+  });
+
+  p = p.then(function (secret) {
+    that.reportSecurityTokenRetreived({securityToken: token});
```

Comment:
bad... leaking secret bits... if the worker doesn't manage to delete secret before it crashes for unknown reason... OR someone races with worker to get secret and delete it...